sql/core/src/main/scala/org/apache/spark/sql/execution/python/AggregateInPandasExec.scala

```diff
@@ -0,0 +1,143 @@
+      case Seq(u: PythonUDF) =>
+        val (chained, children) = collectFunctions(u)
+        (ChainedPythonFunctions(chained.funcs ++ Seq(udf.func)), children)
+      case children =>
+        // There should not be any other UDFs, or the children can't be evaluated directly.
+        assert(children.forall(_.find(_.isInstanceOf[PythonUDF]).isEmpty))
+        (ChainedPythonFunctions(Seq(udf.func)), udf.children)
+    }
+  }
+
+  override def requiredChildOrdering: Seq[Seq[SortOrder]] =
+    Seq(groupingExpressions.map(SortOrder(_, Ascending)))
+
+  override protected def doExecute(): RDD[InternalRow] = {
+    val inputRDD = child.execute()
+
+    val bufferSize = inputRDD.conf.getInt("spark.buffer.size", 65536)
+    val reuseWorker = inputRDD.conf.getBoolean("spark.python.worker.reuse", defaultValue = true)
+    val sessionLocalTimeZone = conf.sessionLocalTimeZone
+    val pandasRespectSessionTimeZone = conf.pandasRespectSessionTimeZone
+
+    val (pyFuncs, inputs) = udfExpressions.map(collectFunctions).unzip
+
+    val allInputs = new ArrayBuffer[Expression]
+    val dataTypes = new ArrayBuffer[DataType]
+
+    allInputs.appendAll(groupingExpressions)
+
+    val argOffsets = inputs.map { input =>
+      input.map { e =>
+        allInputs += e
+        dataTypes += e.dataType
+        allInputs.length - 1 - groupingExpressions.length
+      }.toArray
+    }.toArray
+
+    val schema = StructType(dataTypes.zipWithIndex.map { case (dt, i) =>
+      StructField(s"_$i", dt)
+    })
+
+    inputRDD.mapPartitionsInternal { iter =>
+      val grouped = if (groupingExpressions.isEmpty) {
+        Iterator((null, iter))
+      } else {
+        val groupedIter = GroupedIterator(iter, groupingExpressions, child.output)
+
+        val dropGrouping =
+          UnsafeProjection.create(allInputs.drop(groupingExpressions.length), child.output)
+
+        groupedIter.map {
+          case (k, groupedRowIter) => (k, groupedRowIter.map(dropGrouping))
+        }
+      }
+
+      val context = TaskContext.get()
+
+      // The queue used to buffer input rows so we can drain it to
+      // combine input with output from Python.
+      val queue = HybridRowQueue(context.taskMemoryManager(),
+        new File(Utils.getLocalDir(SparkEnv.get.conf)), groupingExpressions.length)
+      context.addTaskCompletionListener { _ =>
+        queue.close()
+      }
+
+      // Add rows to queue to join later with the result.
+      val projectedRowIter = grouped.map { case (groupingKey, rows) =>
+        queue.add(groupingKey.asInstanceOf[UnsafeRow])
+        rows
+      }
+
+      val columnarBatchIter = new ArrowPythonRunner(
+        pyFuncs, bufferSize, reuseWorker,
+        PythonEvalType.SQL_PANDAS_GROUP_AGG_UDF, argOffsets, schema,
+        sessionLocalTimeZone, pandasRespectSessionTimeZone)
+        .compute(projectedRowIter, context.partitionId(), context)
+
+      val joined = new JoinedRow
+      val resultProj = UnsafeProjection.create(output, output)
```

Comment:
We need to handle `resultExpressions` for the following cases:

```python
    def test_result_expressions(self):
        import numpy as np
        from pyspark.sql.functions import mean, pandas_udf, PandasUDFType

        df = self.data

        @pandas_udf('double', PandasUDFType.GROUP_AGG)
        def mean_udf(v, w):
            return np.average(v, weights=w)

        result1 = (df.groupby('id')
                   .agg(mean_udf(df.v, lit(1.0)) + 1)
                   .sort('id')
                   .toPandas())

        expected1 = (df.groupby('id')
                     .agg(mean(df.v) + 1)
                     .sort('id')
                     .toPandas())

        self.assertPandasEqual(expected1, result1)
```

Reply:
Thanks @ueshin for reminding me of this. Just want to clarify the semantics:

Does
```
 .agg(mean(df.v) + 1)
```
mean "compute mean of df.v and plus the mean by one as output", i.e, same as

```
.agg(mean(df.v).alias('mean'))
.withColumn('mean', col('mean') + 1)
```
?

Reply:
Yes, I think so about the behavior. I guess the plan could be different, though.
We can compare the behavior with non-udf aggregation and let's follow the behavior.

Reply:
I added `ExtractGroupAggPandasUDFFromAggregate`